### PR TITLE
implementation of runtime tables

### PR DESCRIPTION
**Pull request by mimoo**
This PR implements runtime tables, following [RFC: Extended lookup tables](https://github.com/o1-labs/proof-systems/pull/250). 

## Overview

IMG_B6603DA52925-1

## Implementation details

**setup**:
- [x] we obtain the runtime table configuration from the caller, which consists of table IDs (~must be negative~) and length
- [x] we integrate that in both the big concatenated table (an indexing on the first column that restarts from 0 for each table) and the table id vector (now contains the IDs of the runtime tables)
- [x] we also want to keep track of where the runtime table applies (we pre-compute a runtime selector, that will live in both the prover and verifier index) as well as what the offset row of the runtime tables in the big concatenated table is (will be useful later)
- [x] the linearization must add a new constraint if runtime tables are used: `RT(x) (1 - selector_RT(x)) = 0` to enforce that the runtime table `RT` in the proof only has values where it is allowed to have them (enforced by `selector_RT`)
- [x] - powers of alphas must either handle this branch, or add one more constraint always
- [x] the linearization must also modify the lookup constraint if runtime tables are used, to replace the `t` term of the nominator with `t + runtime_stuff` where `runtime_stuff = j * RT(x) + \beta * j * RT(x \omega)`
- [x] add the relevant commitments to the linearization columns (in `linearization_columns`)

**prover side**: 
- [x] we obtain the actual runtime tables from the caller (as a list of ~negative~ table IDs, and values that must match the configuration)
- [x] from these runtime tables, we can create the runtime polynomial that evaluates to all these values (or 0 when not applied) and save a commitment to it in the proof
- [x] we must add zero-knowledge to this runtime table, we can take advantage of the last `ZK_ROWS` of the domain that are already made for that
- [x] we later evaluate this runtime table at the two challenge points and save them in the proof (as well as absorb them)
- [x] we also produce an evaluation proof for these two evaluations
- [x] since `f` is built using the linearization, we don't have to do anything but pass what the environment needs
- [x] since `aggreg` and `sorted` are computed manually, we modify their behavior so that when runtime tables are used the big concatenated table used also contains the contribution/mask from the runtime table
- [x] since the quotient poly `t` is computed via the linearization, we don't have to do anything either

**verifier side**:
- [x] we use the constraints (linearization) all the time, so this will combine the runtime table with the table for us
- [x] we add the evaluation of the runtime table vector to the aggregated proof
- [x] we also need to absorb the runtime commitment (and evaluations) at the correct places

**ocaml**:
- [x] update associated `Caml` types with new fields

**tests**:
- [x] write a simple test to see if we can handle multiple runtime tables

## Data structure changes

proof:

```diff
pub struct LookupCommitments<G: AffineCurve> {
     // ...
+    /// Optional Commitment to concatenated runtime tables
+    pub runtime: Option<PolyComm<G>>,
}

pub struct LookupEvaluations<Field> {
    // ...
+    /// Optionally, a runtime table polynomial
+    pub runtime: Option<Field>,
}
```

prover index:

```diff
pub struct LookupConstraintSystem<F: FftField> {
    // ...
+    /// An optional runtime table selector. It is 0 everywhere,
+    /// except at the rows where the runtime tables apply.
+    pub runtime_selector: Option<E<F, D<F>>>,
    // ...
}

pub struct LookupConfiguration<F: FftField> {
    // ...
+    /// The runtime tables used,
+    /// seen as tuples of `(length, id)`.
+    pub runtime_tables: Option<Vec<RuntimeTableConfiguration>>,
+
+    /// The offset of the runtime table within the concatenated table
+    pub runtime_table_offset: Option<usize>,
    // ...
}
```

verifier index:

```diff
pub struct LookupVerifierIndex<G: CommitmentCurve> {
    // ...
+    /// An optional selector polynomial for runtime tables
+    pub runtime_tables_selector: Option<PolyComm<G>>,
}
```

## Notes 

From discussion with matthew on things we want to do next:

- [ ] make it work with arbitrary-length runtime tables
- [ ] we should have a mode where indexing of these tables can be provided by the caller/front-end as well (not just a counter starting from 0)
- [ ] optimization: in the constraint`(rt_selector - 1) * rt_table` we can precompute `one_min_rt_selector` as `rt_selector -1` and use that in the index
- [x] let's not enforce negative IDs (I'm not totally a fan of this, but let's add a comment for now warning the caller)

### CODE REVIEW

**Comment by mimoo** [comment]
one thing I'm still missing: 

- [ ] the runtime selector should actually not apply after the runtime tables (right now it applies on the dummy rows). But it should still apply on the last ZK_ROWS

**Comment by mimoo** [comment]
For some reason I was implementing the runtime constraint as 

```
        var(Column::LookupRuntimeTable) * (E::one() - var(Column::LookupRuntimeSelector));
```

when I could just do 

```
        var(Column::LookupRuntimeTable) * var(Column::LookupRuntimeSelector);
```

File: kimchi/src/circuits/lookup/runtime_tables.rs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    expr::{prologue::*, Column},
+    gate::CurrOrNext,
+};
+use ark_ff::Field;
+use serde::{Deserialize, Serialize};
+
+/// The configuration of a runtime table.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct RuntimeTableConfiguration {
+    /// The table id (must be negative).
```
[the remark this answers is not in the document]
Comment (mimoo) [reply kimchi/src/circuits/lookup/runtime_tables.rs:13]: changed to just mention that we shouldn't collide

File: kimchi/src/circuits/lookup/index.rs [hunk 2]
```diff
@@ -85,11 +95,63 @@ impl<F: FftField + SquareRootField> LookupConstraintSystem<F> {
...
+                        }
+
+                        // compute the length of the runtime table
+                        let mut runtime_len = 0;
+                        for t in runtime_tables {
+                            runtime_len += t.len;
+                        }
+
+                        // compute the runtime selector
+                        let runtime_selector = {
```
[the remark this answers is not in the document]
Comment (mimoo) [reply kimchi/src/circuits/lookup/index.rs:119]: thinking more about it, maybe we should call it something else than selector? To me selector is what we used to select gates on different rows, whereas this is something else. Perhaps `runtime_active_rows`?

**Comment by mimoo** [comment]
ping @mrmr1993 , merged master and cleaned comments

**Review by mrmr1993 — APPROVED** [review]

**Pull request merged**